### PR TITLE
Allow the Paypal IPN to work with Drupal 8

### PR DESCRIPTION
**Pull request by dsnopek**
Overview
----------------------------------------
Currently, the ipn.php script doesn't work on Drupal 8.

Before
----------------------------------------
PHP Fatal Error

After
----------------------------------------
Appears to work as expected :-)

Technical Details
----------------------------------------
The main issue is that Drupal isn't bootstrapped. However, that isn't totally enough, because there's a permission check that somehow manages to happen before the Drupal container can get set on the `\Drupal` static class. The permission check isn't useful anyway in this situation, and it would probably be nice in general if `CRM_Core_Permission_Drupal8::check()` wouldn't throw an exception when called when Drupal is not ready.

### CODE REVIEW

**Comment by MegaphoneJon** [comment]
This looks like it might also be the technique to solve the Drupal8 CKEditor image uploading issue.  I'm going to take a stab at that.

**Comment by eileenmcnaughton** [comment]
@dsnopek Im confused because the extern/ipn.php is a legacy ipn call  & new sites should use

civicrm/payment/ipn/x where x is the payment processor id - wouldn't all d8 sites be new?

**Comment by eileenmcnaughton** [comment]
@MegaphoneJon I did try to solve the kcfinder & push it through civi routing here https://github.com/eileenmcnaughton/civicrm-core/commit/84848fe055b53c839d4f9426aad2b2321ade84f4 - it's an issue in all cms under some circumstances - but I didn't nail it :-(

**Comment by dsnopek** [comment]
> @dsnopek Im confused because the extern/ipn.php is a legacy ipn call & new sites should use
> 
> civicrm/payment/ipn/x where x is the payment processor id - wouldn't all d8 sites be new?

Ah, hrm. I guess I didn't know that. :-) I setup the site following this documentation:

https://docs.civicrm.org/sysadmin/en/latest/setup/payment-processors/paypal-pro/#recurring-contributions

If the new IPN URL is documented somewhere, I missed it.

**Comment by eileenmcnaughton** [comment]
@dsnopek hmm I meant to open a PR to change it but I think it just committed straight away

**Comment by dsnopek** [comment]
@eileenmcnaughton The URL you put in the docs isn't right, I think. For Drupal it gives:

```
http://example.org/sites/all/modules/civicrm/payment/ipn/x
```

Where I think for all CMS's it should be:

```
http://example.org/civicrm/payment/ipn/x
```

Right?

**Comment by eileenmcnaughton** [comment]
ug - you are right @dsnopek  - actually for Joomla! I'm not quite sure

File: CRM/Core/Permission/Drupal8.php [hunk 1]
```diff
@@ -48,6 +48,10 @@ class CRM_Core_Permission_Drupal8 extends CRM_Core_Permission_DrupalBase {
    * @return bool
    */
   public function check($str, $userId = NULL) {
+    if (!\Drupal::hasContainer()) {
+      return FALSE;
+    }
```
Comment (monishdeb) [inline CRM/Core/Permission/Drupal8.php]: or 
```php
if (!class_exists('Drupal')) {
      return NULL;
    }
```
@dsnopek ?

as I recall in past I hit with an error as ``` unknown class Drupal``` when a core function is invoked before bootstrap.

**Comment by mattwire** [comment]
@dsnopek @eileenmcnaughton @monishdeb @seamuslee001 The issue of payment processor IPN URL keeps on coming up... so I've just taken a few minutes to update the sysadmin docs properly - could you review https://github.com/civicrm/civicrm-sysadmin-guide/pull/152 please and let's get it merged so we don't keep getting questions about which are the correct webhook URLs for each platform?

File: CRM/Core/Permission/Drupal8.php (hunk 1 above)
Comment (dsnopek) [reply CRM/Core/Permission/Drupal8.php]: In the error I was hitting, the `\Drupal` class existed, it just hadn't had `\Drupal::setContainer()` called on it yet. So, the autoloading was setup, but the bootstrap not completed yet.

**Pull request closed without merging**